### ISSUE TITLE
Invalid triggers when using let expressions

### ISSUE THREAD

**Issue by jamesbornholt** [Microsoft/dafny#167]
These two (simplified) functions, which each define a second collection in terms of a first collection, both produce Boogie translation errors:

```
function Rewrite(env: map<nat, nat>): map<nat, nat> {
    var p := map n: nat | n in env :: n;
    map n: nat | n in p :: n
}

function Rewrite2(strs: set<string>): map<string, string> {
    var p := map s: string | s in strs :: s;
    map s: string | s in p :: s
}
```
> foo.dfy(2,13): Error: quantifiers are not allowed in triggers
> (0,-1): Error: boolean operators are not allowed in triggers
> (0,-1): Error: arithmetic comparisons are not allowed in triggers
> foo.dfy(2,13): Error: quantifiers are not allowed in triggers
> foo.dfy(2,13): Error: quantifiers are not allowed in triggers
> (0,-1): Error: boolean operators are not allowed in triggers
> (0,-1): Error: arithmetic comparisons are not allowed in triggers
> foo.dfy(2,13): Error: quantifiers are not allowed in triggers
> foo.dfy(7,13): Error: quantifiers are not allowed in triggers
> foo.dfy(7,13): Error: boolean operators are not allowed in triggers
> foo.dfy(7,13): Error: quantifiers are not allowed in triggers
> foo.dfy(7,13): Error: quantifiers are not allowed in triggers
> foo.dfy(7,13): Error: boolean operators are not allowed in triggers
> foo.dfy(7,13): Error: quantifiers are not allowed in triggers
> 14 name resolution errors detected in foo__module.bpl
> 
> *** Encountered internal translation error - re-running Boogie to get better debug information
> 
> foo__module.bpl(2627,33): Error: quantifiers are not allowed in triggers
> foo__module.bpl(2627,84): Error: boolean operators are not allowed in triggers
> foo__module.bpl(2627,63): Error: arithmetic comparisons are not allowed in triggers
> foo__module.bpl(2628,17): Error: quantifiers are not allowed in triggers
> foo__module.bpl(2649,33): Error: quantifiers are not allowed in triggers
> foo__module.bpl(2649,84): Error: boolean operators are not allowed in triggers
> foo__module.bpl(2649,63): Error: arithmetic comparisons are not allowed in triggers
> foo__module.bpl(2650,17): Error: quantifiers are not allowed in triggers
> foo__module.bpl(2698,33): Error: quantifiers are not allowed in triggers
> foo__module.bpl(2698,93): Error: boolean operators are not allowed in triggers
> foo__module.bpl(2699,17): Error: quantifiers are not allowed in triggers
> foo__module.bpl(2719,33): Error: quantifiers are not allowed in triggers
> foo__module.bpl(2720,58): Error: boolean operators are not allowed in triggers
> foo__module.bpl(2721,17): Error: quantifiers are not allowed in triggers
> 14 name resolution errors detected in foo__module.bpl

It seems that the selected triggers for the second map in each function somehow have the definitions from the first map substituted in, causing them to be invalid.

**Comment by wilcoxjay** [comment]
This is probably due to inlining before trigger selection.

**Comment by jamesbornholt** [comment]
The selected triggers actually appear to be reasonable enough at the Dafny level, without any signs of inlining (`n in p` is selected for the second line of `Rewrite`). The problem is that the `n in p` trigger translates down to Boogie as something like `Map#Domain(Map#Glue((lambda y :: ...), ..., ...))[n]`, where the body of the lambda contains the domain definition for `p` (i.e., `0 <= n`). 

One solution might just be to not select map operations as triggers (indeed, adding `{:trigger false}` is enough to suppress the problem while proofs involving `Rewrite` still go through), but that might be too conservative.

**Comment by jamesbornholt** [comment]
14584de reverts the fix for this. Now I understand the code better I can see that fix was broken.

We need to do something smarter here. One aggressive option might just be to inline all let-expressions early in the rewriter, but we might have to be careful about site effects in the RHS. Another approach would be to explicitly track the RHS of all in-scope let expressions and use them in trigger selection (so having the same effect as inlining, but only for the purpose of trigger selection). Both are more involved than the original, broken fix.

**Comment by jamesbornholt** [comment]
Here's another example showing the general issue -- let expressions allow us to hide things that would usually be trigger killers:

```
function sum(a: int, b: int): int {
  a + b
}

predicate sum_is_sum(b: int, c: int) {
  var s := a => sum(a, b);
  forall a: int :: s(a) == a + b + c
}
```

(based on [triggers-prevent-some-inlining.dfy](https://github.com/Microsoft/dafny/blob/master/Test/triggers/triggers-prevent-some-inlining.dfy))

**Comment by RustanLeino** [comment]
I looked some more at this problem. I think the right kind of solution is to add let expressions to Boogie and to translate Dafny let expressions into Boogie let expressions. This seems possible, since SMTLIB has featured let expressions for many years.

In the meantime, I implemented a [simple peephole optimization](https://github.com/Microsoft/dafny/commit/118e70f0335bab91ae189d50defa9730f396d6db) that happens to avoid the problem in some very simple cases.
